executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+    /// For example, it could also be binary fixed column, indicating
+    /// where the multiplicity can be non-zero.
+    pub selected_tuple: SelectedExpressions<T>,
+    /// The multiplicity with which the tuple is sent to the bus.
+    /// None if this interaction comes from the RHS of a native lookup,
+    /// as native lookup arguments don't require the multiplicity to be
+    /// committed.
+    /// Note that this is the absolute multiplicity, i.e., for bus receives,
+    /// the negation has been removed.
+    pub multiplicity: Option<AlgebraicExpression<T>>,
+}
+
+impl<T: FieldElement> BusSend<T> {
+    /// Tries to find the matching receive, which should succeed iff. the send has
+    /// a static interaction ID.
+    pub fn try_match_static<'a>(
+        &self,
+        receives: &'a BTreeMap<T, BusReceive<T>>,
+    ) -> Option<&'a BusReceive<T>> {
+        Some(&receives[&self.interaction_id()?])
+    }
+
+    /// Returns the interaction ID if it is a number.
+    /// Sends and receives can be matched by matching the interaction ID.
+    pub fn interaction_id(&self) -> Option<T> {
+        match &self.interaction_id {
+            AlgebraicExpression::Number(id) => Some(*id),
+            _ => None,
+        }
+    }
+}
+
+impl<T: FieldElement> BusReceive<T> {
+    /// Returns whether the receive has an unconstrained multiplicity.
+    /// If it does not, this typically means that the multiplicity is binary constraint.
+    /// Unconstrained receives are equivalent to the RHS of a lookup;
+    /// binary-constrained receives are equivalent to the RHS of a permutation.
+    pub fn is_unconstrained(&self) -> bool {
+        // TODO: This always works in practice, but we should properly check the
+        // range constraints on the multiplicity column.
+        self.multiplicity.as_ref() != Some(&self.selected_tuple.selector)
+    }
+}
+
+impl<T> Children<AlgebraicExpression<T>> for BusSend<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        self.selected_tuple.children_mut()
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        self.selected_tuple.children()
+    }
+}
+
+impl<T> Children<AlgebraicExpression<T>> for BusReceive<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children_mut()
+                .chain(self.multiplicity.iter_mut()),
+        )
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children()
+                .chain(self.multiplicity.iter()),
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusSend<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BusSend(tuple={}, interaction_id={})",
```

Comment:
what about `data` instead of `tuple`? or `values`? (not just wrt display)

Reply:
What's wrong with tuple? We also call it tuple in the PIL std lib, so it would be inconsistent.

Reply:
I know and I also don't like it there. What about `payload`?

Reply:
My main point is that `tuple` says next to nothing. If we group `multiplicity` and `interaction_id` in the `receive` below, it's also a tuple.

Reply:
OK, renamed to `payload`.